### DISCUSSION TITLE
How to Batch Upload FAQ's to Database

### DISCUSSION THREAD

**Discussion by c1972** [thorsten/phpMyFAQ#3019] [Show and tell]
If you get bored typing in all the FAQ's one by one, then use my little guide here and upload a few hundred (or thousand) into the database at once!

### 3 things in advance:
1. You should have a table ready with the data you want to import. The titles of the FAQ's in one column, the HTML-formatted content in the next and the keywords in a third column.
2. You should have access to phpMyAdmin via your web host.
3. You should have already created a few entries by hand in phpMyFAQ.

This guide is intentionally extensive, but the individual steps can be completed quite quickly once you understand them. It is important that you don't make any mistakes, so I prefer to put a little more effort into it. :)

- Open phpMyAdmin
- On the left you will see a list of all the tables in the database (phew, there are a lot!)
- From this list, select the following 4 (one after the other) and click on them once to select them:
> - faqdata
> - faqdata_tags
> - faqcategoryrelations
> - faqdata_user

![liste-tabellen](https://github.com/thorsten/phpMyFAQ/assets/112912128/98b595a8-7e9b-4aaf-b36e-a8133d04cffc)

As soon as you have selected one of the 4 tables, the contents will be displayed on the right.

- Now click on **Export** at the top.
- Select **CSV** as the format
- Make sure that Export **ALL records** is selected
- Click on **Export**
- Save all 4 tables in a separate folder on your computer.

Now you can close phpMyAdmin for the moment.

Open the table **faqdata.csv** with your spreadsheet program (I like to use LibreOffice, but any other will work too!). Take a good look at the table first before you change anything! Pay attention to the column headings and contents. Do **NOT** edit the headings! Remember to work very carefully in the table! The database doesn't like mistakes and quickly reacts allergically.

- The **id** column contains an ascending number that is assigned once and identifies each post.
- So for a new post you only need to count one further.
- The same applies to the **solution_id**.
- The **keywords** for the respective post should be entered in the Keywords column, separated by commas.
- In the **Topic** column you enter the question or the title of the post.

- The HTML-formatted content of the post should be entered in the **Content** column.
- If you haven't had anything to do with HTML before, don't worry: you can do it!
- Take a look at how the existing posts are formatted!
- You can use `<p>Paragraph</p>` to enclose a paragraph. `<br>` is a line break. You can use `<strong>BOLD</strong>` to write something in bold. And over time you will find a whole range of other options!

For the remaining columns, just use the existing entries as a guide. Check everything you have entered again. It's better to look at it once too often at the beginning until it becomes easier for you later. Save the table in exactly the same format as a CSV file, then you can close it.

Now you might be thinking... Great, everything important has been entered... Just upload it and you're done. Well, unfortunately it's not quite that easy. In order for the data from this first table to be displayed, the exact right values ​​in the other three tables are also required. Only then will it work and your data will be displayed.

You've probably already assigned some important **tags** to your existing posts, otherwise please do that quickly before we continue. It's best to just use one tag per post.

So now let's take the second table **faqdata_tags**. Thank goodness it only has 2 columns, how relaxing!

On the left is the **record_id**, which matches the unique id from the previous table, but just is named differently. So for a new post you have to increment one here too. Remember that this number refers to the exact post whose data you have already entered in the faqdata table.

In the right-hand column there is a number that identifies the assigned tag.

So we still need to find out which tag is meant by each number. To do this we go to phpMyFAQ, in the Tags section. In the list of tags there is a blue symbol on the right for editing and a red symbol for deleting. Just hold the mouse over one of the trash can symbols (DO NOT click!), your browser will now show you a link at the bottom left that ends with our number. Aha! So that's where we get the thing from, great!

You should note down the tags and their numbers in a text file, for example.

With this knowledge we can now enter the desired tags in the table for our new posts. Check again and you can close the table.

We continue with the third table, called **faqcategoryrelations**, and this is now about the categories.
The number in the first column **category_id** identifies the category in which the new post should be sorted. You can assign the number to the category in exactly the same way as you did with the tags. So make a quick note in the text file. You can now enter the numbers so that the posts end up in the right category. On the right at record_id we have the unique ID that identifies the post.
Is everything entered? Great! Take another look, save and close.

The fourth one called faqdata_user.csv is easy. Just put in the unique id at the left and -1 at the right. Save and close.

**All respect! You've stuck with it up to now and are still in? GREAT!**

Now we have our 4 tables ready to upload. So we open phpMyAdmin again, where the database is already waiting for us.

- Select the correct table on the left.
- Click on Import at the top.
- Select the correct one of our 4 CSV tables out of your folder.
- For Skip number of queries, select 1 so that our column headings are not created as a post. :D
- The format must be CSV.
- Check also to **update on duplicate key**. Otherwise changes to existing entries won't be saved.
- Otherwise everything should be fine.
- Now click on import and keep your fingers crossed that there is no error message.

We'll do that three times and then **FINALLY!!!** we can quickly check in phpMyFAQ to see if our data has arrived!! Hopefully I haven't forgotten anything!!

**Phew**, that was a long guide!! I hope that lots of people can use it. I'm happy about every comment on how it worked, every thank you and every thumbs up!

Have fun with phpMyFAQ and best wishes, Christian!

**Comment by thorsten** [comment]
Nice summary, but you can use the API to add FAQs automatically. With the upcoming 4.0 release there's a CSV Import functionality as well.

The API documentation (for 4.0) can be found here: https://api-docs.phpmyfaq.de/

The docus about the CVS Import is here: https://phpmyfaq.readthedocs.io/en/main/administration/#541-imports

**Comment by c1972** [comment]
It's a manual for people who usually don't handle databases. This has been asked about few times in the forum.

The CSV import function will distribute the uploaded data across different database tables, right?
You could think it would just end up in ONE table. But those tables stay the same with 4.0, or?

I didn't knew about the API until now (and don't understand how to use it).

**Reply by thorsten** [reply]
> You could think it would just end up in ONE table. But those tables stay the same with 4.0, or?

yes, the table structure in 4.0 won't change much.

If you like you can contribute your documentation to the main documentation: https://github.com/thorsten/phpMyFAQ/tree/main/docs

**Comment by c1972** [comment]
Klar würd ich das gerne in die Dokumentation packen.
Ich kenne mich aber noch nicht so gut aus bei Github und möchte kein Chaos in Euren Daten anrichten.
Kannst Du mir etwas genauer sagen, an welche  Stelle meine Anleitung hin gehört?
Und hier hab ich es teils mit irgendwelchen Sprüchen aufgehübscht, sowas gehört sicher nicht in die Doku, also müsste ich es etwas aufräumen, richtig?

**Comment by c1972** [comment]
4.51 Also unterhalb Add FAQ
oder
Als Teil von 5.22 FAQ Administration
?

Achja, sind da sicher keine Fehler im Umgang mit der DB drin in meinem Text?

**Comment by c1972** [comment]
So, ich hab meinen Text dafür schon verbessert und angepasst.

**Comment by c1972** [comment]
Jetzt Fork und da meine Änderungen rein ja?

**Comment by c1972** [comment]
OK, fertig. Da bin ich jetzt gespannt! :)

**Comment by c1972** [comment]
![comparing-changes](https://github.com/thorsten/phpMyFAQ/assets/112912128/b1009eac-584b-41ff-a2b4-58490f29b11f)

Muss ich da jetzt noch was machen? Oder warte ich schlicht, bis ihr das für ok befunden habt?

**Comment by c1972** [comment]
Pull request hab ich noch gemacht... und nu?? :|
Ich wollte ich wüßte was ich da mache haha :D

**Comment by c1972** [comment]
Hab's jetzt in **administration.md** als Punkt 5.4.3 Batch upload via phpMyAdmin.
Ist das angekommen?
(Sorry, ich brauch ne Zeit bis ich mich hier zurecht finde.)
